metrics/graphite/graphite.go

```diff
@@ -175,18 +175,48 @@ func (e *Emitter) Stop() error {
 	return err
 }
 
+var (
+	RetryMax        = 10
+	RetryWait       = 2 * time.Millisecond
+	RetryMultiplier = 2
+)
+
 // Flush will write the current metrics to the Emitter's
 // connection in the Graphite plaintext protocol.
-func (e *Emitter) Flush() { e.flush(e.conn) }
-
-func (e *Emitter) flush(conn io.Writer) {
+func (e *Emitter) Flush() error {
 	// only one flush at a time
 	e.mtx.Lock()
 	defer e.mtx.Unlock()
 
+	// set the system up to perform a retry loop
+	var err error
+	wait := RetryWait
+	for i := 1; i <= RetryMax; i++ {
+		err = e.flush(e.conn)
+		if err == nil {
+			return nil
+		}
+		e.logger.Log(
+			"err", err,
+			"msg", fmt.Sprintf("unable to flush metrics on attempt %d, waiting %s", i, wait),
+		)
+		time.Sleep(wait)
```

Comment:
Minor point, but the last (and longest) sleep will still occur even though we're not going to retry the flush. You might want to special-case that last one to break immediately after the failed flush.